src/librustc_incremental/persist/fs.rs

```diff
@@ -0,0 +1,810 @@
+            // some kind of invalid directory name
+            continue
+        };
+
+        if is_match {
+            return Some(OsString::from(sub_dir_name))
+        }
+    }
+
+    None
+}
+
+fn crate_path(sess: &Session,
+              crate_name: &str,
+              crate_disambiguator: &str)
+              -> PathBuf {
+    use std::hash::{SipHasher, Hasher, Hash};
+
+    let incr_dir = sess.opts.incremental.as_ref().unwrap().clone();
+
+    // The full crate disambiguator is really long. A hash of it should be
+    // sufficient.
+    let mut hasher = SipHasher::new();
+    crate_disambiguator.hash(&mut hasher);
+
+    let crate_name = format!("{}-{:x}", crate_name, hasher.finish());
+    incr_dir.join(crate_name)
+}
+
+fn assert_no_characters_lost(s: &str) {
+    if s.contains('\u{FFFD}') {
+        bug!("Could not losslessly convert '{}'.", s)
+    }
+}
+
+pub fn garbage_collect_session_directories(sess: &Session) -> io::Result<()> {
+    debug!("garbage_collect_session_directories() - begin");
+
+    let session_directory = sess.incr_comp_session_dir();
+    debug!("garbage_collect_session_directories() - session directory: {}",
+        session_directory.display());
+
+    let crate_directory = session_directory.parent().unwrap();
+    debug!("garbage_collect_session_directories() - crate directory: {}",
+        crate_directory.display());
+
+    let mut deletion_candidates = vec![];
+    let mut definitely_delete = vec![];
+
+    for dir_entry in try!(crate_directory.read_dir()) {
+        let dir_entry = match dir_entry {
+            Ok(dir_entry) => dir_entry,
+            _ => {
+                // Ignore any errors
+                continue
+            }
+        };
+
+        let directory_name = dir_entry.file_name();
+        let directory_name = directory_name.to_string_lossy();
+
+        if !directory_name.starts_with("sess-") {
+            // This is something we don't know, leave it alone...
+            continue
+        }
+
+        assert_no_characters_lost(&directory_name);
+
+        debug!("garbage_collect_session_directories() - inspecting: {}",
+                directory_name);
+
+        match extract_timestamp_from_session_dir(&directory_name) {
+            Ok(timestamp) => {
+                if directory_name.ends_with("-working") {
+                    let ten_seconds = Duration::new(10, 0);
+
+                    // When cleaning out "-working" session directories, i.e.
+                    // session directories that might still be in use by another
+                    // compiler instance, we only look a directories that are
+                    // at least ten seconds old. This is supposed to reduce the
+                    // chance of deleting a directory in the time window where
+                    // the process has allocated the directory but has not yet
+                    // acquired the file-lock on it.
+                    if timestamp < SystemTime::now() - ten_seconds {
```

Comment:
Using `SystemTime::now().duration_since(timestamp) > ten_seconds` is nicer, but may fail (since duration_since could be "negative"). Another alternative is `timestamp + ten_seconds < SystemTime::now()`.